testutil: add DeepUnsortedMatches Checker

MiguelPires opened the pull request:
Adds a Checker to test that two containers have the same elements (duplicates must occur in the same number) but possibly in different orders.

mardy commented on testutil/containschecker.go:
```diff
@@ -150,3 +150,69 @@ func (c *deepContainsChecker) Check(params []interface{}, names []string) (resul
+
+	if container1.Len() != container2.Len() {
+		return false, fmt.Sprintf("containers have different lengths: %d != %d", container1.Len(), container2.Len())
+	}
```
Is there a way to also print the contents of the two containers? That would be helpful in debugging.

MiguelPires replied:
I can change it back to printing the differences only (length, type, index x is unmatched, etc) and

mardy commented on testutil/containschecker.go:
```diff
@@ -150,3 +150,69 @@ func (c *deepContainsChecker) Check(params []interface{}, names []string) (resul
+// the same number (but possibly different order) using DeepEqual. The container
+// can be an array, a slice or a map.
+var DeepContainerMatches check.Checker = &deepContainerMatchChecker{
+	&check.CheckerInfo{Name: "DeepContainerMatchChecker", Params: []string{"container1", "container2"}},
```
I'm not sure about this name: the word "deep" already tells us that this works on containers, so the "container" part seems unnecessary. On the other hand, the most important characteristic of this matcher (the fact that it doesn't care about sorting) is not mentioned in its name. Maybe "DeepUnsortedMatchChecker" could be a better name?

MiguelPires replied:
:+1: I went back and forth on the name but DeepUnsortedMatches and deepUnsortedMatchChecker seems clearer. I'll change it, thanks

mardy replied:
Thanks!

mardy commented on testutil/containschecker_test.go:
```diff
@@ -221,3 +221,72 @@ func (*containsCheckerSuite) TestDeepContainsUncomparableType(c *check.C) {
+	}
+
+	c.Assert(slice1, DeepContainerMatches, slice2)
+	c.Assert(slice2, DeepContainerMatches, slice1)
```
nitpick: I would vote for using `Check` here instead of `Assert`.

mardy commented on testutil/containschecker_test.go:
```diff
@@ -221,3 +221,72 @@ func (*containsCheckerSuite) TestDeepContainsUncomparableType(c *check.C) {
+func (*containsCheckerSuite) TestDeepContainerMatchesSliceSuccess(c *check.C) {
+	slice1 := []dummy{
+		{a: "one", b: map[string]int{"a": 1}},
+		{b: map[string]int{"b": 2}, a: "two"},
```
Isn't the order of the initializers irrelevant in any case? I think that swapping the order makes this test harder to read and risks letting the reader miss the point of the test. I'd always initialize first `a` then `b`, and just swap the order of the elements between `slice1` and `slice2`.

mardy reviewed (commented):
A few suggestions, as usual only about naming and readability :-)

MiguelPires commented:
Thanks for the feedback @mardy. I improved the name and error messages based on your comments. My concern was that the error output would be hard to read for more complex structures  but I guess we don't lose anything by doing it and indeed it's very useful for the simpler stuff

anonymouse64 commented on testutil/containschecker.go:
```diff
@@ -150,3 +150,73 @@ func (c *deepContainsChecker) Check(params []interface{}, names []string) (resul
+		for e := 0; e < container2.Len(); e++ {
+			el2 := container2.Index(e).Interface()
+
+			if !matched[e] && reflect.DeepEqual(el1, el2) {
+				matched[e] = true
```
can you add some comments to this code? how I think this code is working (please correct me if I'm wrong) is that we essentially take an element in container1, match it to something in container2 and mark the index we matched it with, so that we can't ever match multiple items in container1 with the same index in container2. But it took me a while to figure that out while staring at the code, so having a comment somewhere in the function would be good to folks like myself :smile:

MiguelPires replied:
I'll add a comment but just wanted to say your description is exactly right. Marking each index as already matches prevents duplicate elements in container1 from being matched to the same element in container2 (which could then have elements not in container1 as long as the length was the same, e.g., [a, a, b] [a b c]).

anonymouse64 commented on testutil/containschecker_test.go:
```diff
@@ -221,3 +221,72 @@ func (*containsCheckerSuite) TestDeepContainsUncomparableType(c *check.C) {
+}
+
+func (*containsCheckerSuite) TestDeepUnsortedMatchesSliceFailure(c *check.C) {
+	slice1 := []string{"a", "a", "b"}
+	slice2 := []string{"b", "a", "c"}
```
we also should have some positive test cases like:

```go
c.Check([]string{"a","a"}, DeepUnsortedMatches, []string{"a","a"}) // only duplicates but same number
c.Check([]string{"a","b", "a"}, DeepUnsortedMatches, []string{"b", "a","a"}) // duplicates but also other things
c.Check([]string{"a","b", "a"}, DeepUnsortedMatches, []string{"a", "a","a"}) // duplicates but doesn't match
```

anonymouse64 reviewed (commented):
Code is good, thanks for adding this checker, there are multiple places in our codebase where we sort things that don't need to be sorted precisely because we lacked this sort of checker and (speaking for myself at least) we were too lazy to implement this checker.

I do think we need some more tests for this function though, and more comments in the main implementation would be great

MiguelPires commented:
Thanks for your feedback. I changed the errors messages back to only print differences (length/kind is difference, element x is unmatched, etc). I think in most cases we'll use this to compare slices and maps of structs so it might be better to leave it to each person to decide what to print. I also added a few more tests cases and improved the handling of nil values (a lot of the reflection methods panic if called on the Invalid Value so the Checker is now more careful with that). One thing that this doesn't do is deal with pointers to containers gracefully, it will just complain about the type being unsupported (I added a test for that)

mardy reviewed (approved):
LGTM!

anonymouse64 reviewed (approved):
LGTM, thanks for the updates
